### PR TITLE
ENYO-1998: Add supports for subscribed requests and cancel function

### PR DESCRIPTION
**Pull request by viodragon2**
Issue
---
- Reference to subscribed request object is not preserved anywhere, making it garbage collected and lose subscribed callbacks
- There's no way to cancel `LS2Request` with `LunaSource`

Fix
---
- Keep active requests and active subscribed requests references in `LunaSource`
- Create cancel function in LunaSource

Enyo-DCO-1.1-Signed-off-by: Stephen Choi <stephen.choi@lge.com>

### CODE REVIEW

File: src/LunaSource.js [hunk 1]
```diff
@@ -70,19 +83,63 @@ module.exports = kind(
...
+		return model.request;
+	},
+
+	cancel: function (req) {
+		this.removeRequest(req);
+		req.originalCancel();
 	},
 
+	removeRequest: function (req) {
+		var i = -1;
```
Comment (webOS101) [inline src/LunaSource.js]: No need to initialize if you set it immediately.

File: src/LunaSource.js [hunk 2]
```diff
@@ -70,19 +83,63 @@ module.exports = kind(
...
+	requestFailure: function (opts, req, error) {
+		if(opts.error) {
+			opts.error(error, req);
+		}
+
+		this.requestComplete(req);
+	},
+
+	requestComplete: function (req) {
+		var i = -1;
```
Comment (webOS101) [inline src/LunaSource.js]: Isn't this just removeRequest?

File: src/LunaSource.js [hunk 3]
```diff
@@ -100,6 +157,17 @@ module.exports = kind(
...
+		var i;
+		for(i=0; i<this.activeRequests.length; i++) {
+			this.activeRequests[i].originalCancel();
+		}
+		delete this.activeRequests;
+
+		for(i=0; i<this.activeSubscriptionRequests.length; i++) {
+			this.activeSubscriptionRequests[i].originalCancel();
+		}
+		delete this.activeSubscriptionRequests;
```
Comment (webOS101) [inline src/LunaSource.js]: I think it's better to just set these to empty arrays.

File: src/LunaSource.js [hunk 4]
```diff
@@ -51,6 +52,18 @@ module.exports = kind(
 	kind: Source,
 
 	/**
+	*
+	* @private
+	*/
+	activeRequests: [],
```
Comment (webOS101) [inline src/LunaSource.js]: You need to initialize these in create or all will share the same array

File: src/LunaSource.js [hunk 5]
```diff
@@ -70,19 +83,63 @@ module.exports = kind(
...
+			}
+		}
+	},
+
+	requestSuccess: function (opts, req, res) {
+		if(opts.success) {
+			opts.success(res, req);
+		}
+
+		this.requestComplete(req);
```
Comment (webOS101) [inline src/LunaSource.js]: Aren't subscription requests -not- complete?

File: src/LunaSource.js [hunk 6]
```diff
@@ -70,19 +83,63 @@ module.exports = kind(
...
+	},
+
+	requestFailure: function (opts, req, error) {
+		if(opts.error) {
+			opts.error(error, req);
+		}
+
+		this.requestComplete(req);
```
Comment (webOS101) [inline src/LunaSource.js]: This would never remove subscriptions that fail.

File: src/LunaSource.js [hunk 7]
```diff
@@ -67,20 +80,64 @@ module.exports = kind(
...
+
+	cancel: function (req) {
+		this.removeRequest(req);
+		req.originalCancel();
+	},
+
+	removeRequest: function (req) {
+		var i;
+		if (req.subscribe) {
+			i = this.activeSubscriptionRequests.indexOf(req);
```
Comment (aarontam) [inline src/LunaSource.js]: Let's use `utils#remove` here.

File: src/LunaSource.js [hunk 8]
```diff
@@ -67,20 +80,64 @@ module.exports = kind(
...
+			if (i !== -1) {
+				this.activeSubscriptionRequests.splice(i, 1);
```
Comment (aarontam) [inline src/LunaSource.js]: Same as above, let's use `utils#remove` here.

File: src/LunaSource.js [hunk 9]
```diff
@@ -67,20 +80,64 @@ module.exports = kind(
...
 			}
-		});
-		model.request.error(function (req, res) {
-			if(opts.error) {
-				opts.error(res, req);
+		} else {
+			i = this.activeRequests.indexOf(req);
```
Comment (aarontam) [inline src/LunaSource.js]: Same as above, let's use utils#remove here.

File: src/LunaSource.js [hunk 10]
```diff
@@ -67,20 +80,64 @@ module.exports = kind(
...
+	},
+
+	cancel: function (req) {
+		this.removeRequest(req);
+		req.originalCancel();
+	},
+
+	removeRequest: function (req) {
+		var i;
+		if (req.subscribe) {
```
Comment (aarontam) [inline src/LunaSource.js]: Do we need to account for mock requests here, as well?

**Pull request closed without merging**